src/ProjectOrigin.VerifiableEventStore/Services/Repository/Postgres/PostgresqlRepository.cs

```diff
@@ -0,0 +1,224 @@
+                "SELECT transaction_hash, stream_id, stream_index, payload FROM transactions WHERE @fromTransaction <= id AND id <= @toTransaction ORDER BY ID ASC",
+                new { fromTransaction, toTransaction })).ToList();
+
+            var merkleRootHash = transactions.CalculateMerkleRoot(x => x.Payload);
+            var previousHeaderHash = previousBlock?.BlockHash ?? new byte[32];
+            var previousPublicationHash = previousBlock is not null ? SHA256.HashData(previousBlock!.Publication!) : new byte[32];
+
+            var blockHeader = new ImmutableLog.V1.BlockHeader
+            {
+                PreviousHeaderHash = ByteString.CopyFrom(previousHeaderHash),
+                PreviousPublicationHash = ByteString.CopyFrom(previousPublicationHash),
+                MerkleRootHash = ByteString.CopyFrom(merkleRootHash),
+                CreatedAt = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow)
+            };
+
+            await connection.ExecuteAsync(
+                "INSERT INTO blocks(block_hash, previous_header_hash, previous_publication_hash, merkle_root_hash, from_transaction, to_transaction) VALUES (@blockHash, @previousHeaderHash, @previousPublicationHash, @merkleRootHash, @fromTransaction, @toTransaction)",
+                new
+                {
+                    blockHash = SHA256.HashData(blockHeader.ToByteArray()),
+                    previousHeaderHash,
+                    previousPublicationHash,
+                    merkleRootHash,
+                    fromTransaction,
+                    toTransaction
+                });
+
+            await transaction.CommitAsync();
+
+            return new(blockHeader, transactions.Select(x => x.TransactionHash).ToList());
+        }
+        catch
+        {
+            await transaction.RollbackAsync();
+            throw;
+        }
+    }
+
+    public async Task FinalizeBlock(BlockHash hash, ImmutableLog.V1.BlockPublication publication)
+    {
+        using var connection = _dataSource.CreateConnection();
+
+        // update block with publication and return number of rows affected
+        var affectedRows = await connection.ExecuteAsync(
+            "UPDATE blocks SET publication = @publication WHERE block_hash = @blockHash AND (publication IS NULL OR publication = @publication)",
+            new { blockHash = hash.Data, publication = publication.ToByteArray() });
+
+        if (affectedRows == 0)
+            throw new InvalidOperationException("Block not found or already published or publication does not match");
+    }
+
+    private Task<long?> GetTransactionId(IDbConnection connection, TransactionHash transactionHash)
+    {
+        return connection.QuerySingleOrDefaultAsync<long?>(
+              "SELECT id FROM transactions WHERE transaction_hash = @transactionHash",
+              new { transactionHash = transactionHash.Data });
+    }
+
+    private record DBBlock
```

Comment:
It is a bit strange that `DBBlock` is private and at the same time is something that can be fetched from the database.

Reply:
changed name